torch/_mock_dispatcher/dispatch_key_extractor.py

```diff
@@ -1,21 +1,72 @@
+    # since there are a few ops that get fallthrough kernels registered separately for cpu vs cuda.
     nonFallthroughKeys_: DispatchKeySet
+    # Invariant: if requiresBitsetPerBackend_ is True, then we can use nonFallthroughKeys_
+    # to perform the runtime fallthrough key masking.
+    # If it's false, then requiresBitsetPerBackend_ is in an invalid state.
+    requiresBitsetPerBackend_: bool
+    nonFallthroughKeysPerBackend_: List[DispatchKeySet]
     # missing dispatch_arg_indices_reverse
 
     def __init__(self) -> None:
+        # Every backend gets its own fallthrough key bitset.
+        # This is because different backends can independently choose to fallthrough for a given piece of functionality,
+        # E.g. AutogradCPU vs AutogradCUDA
+        self.nonFallthroughKeysPerBackend_ = [DispatchKeySet(full=True) for _ in range(NUM_BACKENDS)]
         self.nonFallthroughKeys_ = DispatchKeySet(full=True)
+        self.requiresBitsetPerBackend_ = False
 
     def setOperatorHasFallthroughForKey(self, k: DispatchKey, has_fallthrough: bool) -> None:
+        # (1) Update nonFallthroughKeys_
         if has_fallthrough:
-            self.nonFallthroughKeys_ = self.nonFallthroughKeys_.remove(k)
+            self.nonFallthroughKeys_ = self.nonFallthroughKeys_.removeFunctionalityBit(k)
         else:
             self.nonFallthroughKeys_ = self.nonFallthroughKeys_.add(k)
 
+        # (2) Update the per-backend nonfallthrough keys.
+        if k != DispatchKey.Undefined and isPerBackendFunctionalityKey(toFunctionalityKey(k)):
+            # This is a per-backend functionality key.
+            # We need to figure out what the currenty backend is,
+            # and only update the bitset for that backend.
+            idx = toBackendKey(k).value
+            if has_fallthrough:
+                self.nonFallthroughKeysPerBackend_[idx] = self.nonFallthroughKeysPerBackend_[idx].removeFunctionalityBit(k)
+            else:
+                self.nonFallthroughKeysPerBackend_[idx] = self.nonFallthroughKeysPerBackend_[idx].add(k)
+        else:
+            # Otherwise, if a fallthrough is set for a functionality that isn't per backend,
+            # Then we update the fallthrough bitset for EVERY backend.
+            for i in range(NUM_BACKENDS):
+                if has_fallthrough:
+                    self.nonFallthroughKeysPerBackend_[i] = self.nonFallthroughKeysPerBackend_[i].removeFunctionalityBit(k)
+                else:
+                    self.nonFallthroughKeysPerBackend_[i] = self.nonFallthroughKeysPerBackend_[i].add(k)
+
+        # (3) set requiresBitsetPerBackend accordingly
+        for i in range(NUM_BACKENDS - 1):
+            if self.nonFallthroughKeysPerBackend_[i] != self.nonFallthroughKeysPerBackend_[i + 1]:
+                self.requiresBitsetPerBackend_ = True
+                return
+        self.requiresBitsetPerBackend_ = False
+
     def getDispatchKeySetUnboxed(self, args: List[DispatchKeySet]) -> DispatchKeySet:
```

Comment:
I added @bhosmer's optimization here: for ops that don't have backend-dependent fall through keys, we can use a single dispatch keyset instead of having to grab the correct one from a list at runtime. This change:

(-) Adds a (very predictable) branch in the hotpath
(+) Avoids a load from a global array of bitsets, when the correct branch is taken

Right now, the good path will be taken for every aten op, since none of Autograd/Dense/Sparse/Quantized register explicit fallthroughs to specific aten ops. If we make Autocast a per-backend key at some point, then a few ops would end up taking the slower path, since according to @ngimel we register a different set of fall through kernels to AutocastCPU/AutocastCUDA.